### PR TITLE
Add Indexed completionMode to Job API

### PR DESCRIPTION
**Pull request by alculquicondor**
**What type of PR is this?**

/kind feature
/kind api-change

/sig apps
/area workload-api/job
/area batch

**What this PR does / why we need it**:

Add the `.spec.completionMode` field with accepted values `NonIndexed` (default) and `Indexed`.

Also requires a feature gate `IndexedJob` and the `.status.completedIndexes` field in Job.

**Which issue(s) this PR fixes**:

Ref #97169 #14188 kubernetes/enhancements#2214

**Special notes for your reviewer**:

This PR only contains API changes. Changes to Job controller will follow up in another PR.

**Does this PR introduce a user-facing change?**:

```release-note
Added `.spec.completionMode` field to Job, with accepted values `NonIndexed` (default) and `Indexed`. This is an alpha field and is only honored by servers with the `IndexedJob` feature gate enabled.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: http://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job
- [Usage]: TODO
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
/hold
for kubernetes/enhancements#2245 to be merged

**Comment by alculquicondor** [comment]
/assign @soltysh 
for sig-apps review

**Comment by alculquicondor** [comment]
Keeping the hold until implementation PR (TODO) is also approved

**Comment by alculquicondor** [comment]
cc @ahg-g

File: pkg/apis/batch/types.go
Comment (ahg-g) [inline pkg/apis/batch/types.go:97]: are there restrictions on what values `Parallelism` and `Completions` can take  when using this mode?

File: pkg/apis/batch/types.go
Comment (ahg-g) [inline pkg/apis/batch/types.go:228]: mention the limit on this field if there is one.

File: pkg/apis/batch/types.go
Comment (ahg-g) [inline pkg/apis/batch/types.go:183]: why not reject it at validation?

File: pkg/apis/batch/validation/validation.go
Comment (ahg-g) [inline pkg/apis/batch/validation/validation.go:34]: add a comment explaining why parallelism is now capped.

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:91]: Any suggested wording?

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:97]: Yes, but I'm keeping them in the field, as that's where people will take a look first.

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:183]: We do reject it at validation. But an apiserver can temporarily be at a newer version than the controller. Then the controller might see a mode that it doesn't recognize.

This note is a safeguard to allow the addition of new modes in the future. In production clusters, the situation described wouldn't happen, as we first release new modes with a disabled feture gate.

But actually, it's safer to skip the sync. Changing.

File: pkg/apis/batch/types.go
Comment (alculquicondor) [reply pkg/apis/batch/types.go:228]: There is no limit here. It's just bounded by parallelism.

File: pkg/apis/batch/validation/validation.go
Comment (alculquicondor) [reply pkg/apis/batch/validation/validation.go:34]: Done

File: pkg/apis/batch/types.go
Comment (ahg-g) [reply pkg/apis/batch/types.go:97]: may be not in this case, but a field can exist in multiple places, and so we wouldn't repeat the description of each value on each field. The documentation is for the type, not really for the field.

File: pkg/apis/batch/types.go
Comment (ahg-g) [reply pkg/apis/batch/types.go:91]: Perhaps:

NonIndexedMode: a job is completed when the number of successfully completed pods is equal to `Completions`, if completions is nil, then the success of any pod signals the success of the job.

IndexedMode: a job is completed when at least one pod for each unique index between 0 to `Completions - 1` successfully completed.

When we start describing the mode from the perspective of the job, I think it becomes more important to document the restrictions on Completions/Parallelism values here.

Comment (alculquicondor) [reply pkg/apis/batch/types.go:91]: Sadly, comments for constants don't make it to the reference documentation https://kubernetes.io/docs/reference/kubernetes-api/

So leaving the comments here lean and putting all details in the field.

Comment (ahg-g) [reply pkg/apis/batch/types.go:91]: oh, that is bad!

**Comment by ahg-g** [comment]
looks good to me

**Review by soltysh — APPROVED** [review]
overall this lgtm from sig-apps pov, I'd suggest you to split manual changes from generated into separate commits when updating this PR
/triage accepted
/priority backlog
/hold cancel
since the enhancement merged already

**Review by erictune — APPROVED** [review]

**Comment by lavalamp** [comment]
/approve
for pkg/registry

**Pull request merged**
